src/Bicep.Core/SemanticModel/NameBindingVisitor.cs

```diff
@@ -110,30 +131,61 @@ private Symbol ValidateFunctionFlags(Symbol symbol, TextSpan span)
             return symbol;
         }
 
-        private Symbol LookupSymbolByName(string name, TextSpan span)
+        private Symbol LookupSymbolByName(string name, TextSpan span, string? @namespace)
         {
-            if (this.declarations.TryGetValue(name, out var localSymbol))
+            NamespaceSymbol? FindNamespace(string name) =>
+                this.namespaces
+                .Where(ns => ns.Name.Equals(name, LanguageConstants.IdentifierComparison))
+                .FirstOrDefault();
+
+            Symbol? foundSymbol;
+            if (string.IsNullOrEmpty(@namespace))
             {
-                // we found the symbol in the local namespace
-                return ValidateFunctionFlags(localSymbol, span);
+                // attempt to find name in the imported namespaces
+                var namespaceSymbol = FindNamespace(name);
+
+                if (namespaceSymbol != null)
+                {
+                    // namespace symbol found
+                    return ValidateFunctionFlags(namespaceSymbol, span);
+                }
+
+                // declarations must not have a namespace value, namespaces are used to fully qualify a function access.
+                // There might be instances where a variable declaration for example uses the same name as one of the imported
+                // functions, in this case to differentiate a variable declaration vs a function access we check the namespace value,
+                // the former case must have an empty namespace value whereas the latter will have a namespace value.
+                if (this.declarations.TryGetValue(name, out var localSymbol))
+                {
+                    // we found the symbol in the local namespace
+                    return ValidateFunctionFlags(localSymbol, span);
+                }
+
+                // attempt to find function in all imported namespaces
+                var foundSymbols = this.namespaces
+                    .Select(ns => ns.TryGetFunctionSymbol(name))
+                    .Where(symbol => symbol != null)
+                    .ToList();
+
+                if (foundSymbols.Count() > 1)
+                {
+                    // ambiguous symbol
+                    return new ErrorSymbol(DiagnosticBuilder.ForPosition(span).AmbiguousSymbolReference(name, this.namespaces.Select(ns => ns.Name)));
+                }
+
+                foundSymbol = foundSymbols.FirstOrDefault();
             }
+            else
+            {
+                // attempt to find name in the imported namespaces
+                var namespaceSymbol = FindNamespace(@namespace!);
```

Comment:
As a general rule, avoid using `!` unless 100% necessary as you're overruling the compiler's type-safety. In this case, you should be able to avoid it by doing the explicit null check (see other comment)

Reply:
+1

The reason for this is that there are no warnings/errors for overuse of `!`, so refactors that modify nullability will keep the override around even if it's not needed. The end result is that you lose the benefit of compile-time null checking. I've gotten into the habit of rewriting my code just to avoid using `!` with the exception of test code. (Although we probably should hold test code to the same standard, but that's a separate issue :))